src/phimv.jl

```diff
@@ -0,0 +1,178 @@
+    sparse problems. The matrix under consideration interacts only
+    via matrix-vector products (matrix-free method).
+
+    Roger B. Sidje (rbs@maths.uq.edu.au)
+    EXPOKIT: Software Package for Computing Matrix Exponentials.
+    ACM - Transactions On Mathematical Software, 24(1):130-156, 1998
+"""
+function phimv{T}( t::Number,
+                   A, u::Vector{T}, vec::Vector{T};
+                   tol::Real=1e-7,
+                   m::Int=min(30, size(A, 1)),
+                   norm=Base.norm)
+    result = convert(Vector{promote_type(eltype(A), T, typeof(t))}, copy(vec))
+    phimv!(t, A, u, result; tol=tol, m=m, norm=norm)
+    return result
+end
+
+phimv!{T}( t::Number,
+           A, u::Vector{T}, vec::Vector{T};
+           tol::Real=1e-7,
+           m::Int=min(30, size(A, 1)),
+           norm=Base.norm) = phimv!(vec, t, A, u, vec; tol=tol, m=m, norm=norm)
+
+function phimv!{T}( w::Vector{T}, t::Number, A, u::Vector{T}, vec::Vector{T};
+                    tol::Real=1e-7, m::Int=min(30, size(A, 1)), norm=Base.norm)
+
+    if size(vec, 1) != size(A, 2)
+        error("dimension mismatch")
+    end
+
+    # safety factors
+    gamma = 0.9
+    delta = 1.2
+
+    btol = 1e-7     # tolerance for "happy-breakdown"
+    maxiter = 10    # max number of time-step refinements
+
+    anorm = norm(A, Inf)
+    rndoff = anorm*eps()
+
+    # estimate first time-step and round to two significant digits
+    beta = norm(A*vec + u)
+    r = 1/m
+    fact = (((m+1)/exp(1))^(m+1))*sqrt(2*pi*(m+1))
+    tau = (1./anorm)*((fact*tol)/(4.*beta*anorm))^r
+    tau = signif(tau, 2)
+
+    # storage for Krylov subspace vectors
+    vm = Array{typeof(w)}(m+1)
+    for i = 1:m+1
+        vm[i] = similar(w)
+    end
+    hm = zeros(T, m+3, m+3)
+
+    tf = abs(t)
+    tsgn = sign(t)
+    tk = zero(tf)
+
+    copy!(w, vec)
+    p = similar(w)
+    k1 = 3
+    mb = m
+
+    while tk < tf
+        tau = min(tf-tk, tau)
+
+        # Arnoldi procedure
+        scale!(copy!(vm[1], A*w+u), 1/beta)
+
+        for j = 1:m
+            A_mul_B!(p, A, vm[j])
+
+            for i = 1:j
+                hm[i, j] = dot(vm[i], p)
+                p = axpy!(-hm[i,j], vm[i], p)
+            end
+            hm[j+1,j] = norm(p)
+
+            if real(hm[j+1,j]) < btol # happy-breakdown
+                tau = tf - tk
+                k1 = 0
+                mb = j
+                break
+            end
+            scale!(copy!(vm[j+1], p), 1/hm[j+1,j])
+        end
+
+        hm[1, mb+1] = one(T)
+        if k1 != 0
+            hm[m+1,m+2] = one(T); hm[m+2,m+3] = one(T)
+            h = hm[m+1,m]; hm[m+1,m] = zero(T)
+            avnorm = norm(A*vm[m+1])
+        end
+
+        F = similar(hm[1:mb + max(1,k1), 1:mb + max(1,k1)])
+
+        # propagate using adaptive step size
+        iter = 0
+        while iter <= maxiter
+            mx = mb + max(1,k1)
+            F = expm(tsgn*tau*hm[1:mx,1:mx])
+            if k1 == 0
```

Comment:
Here you create a new `F`, dus effectively discarding the one you've just allocated above

Reply:
I guess those are residues from MATLAB. In our `expmv` the respective line reads
```
                F = expm!(tsgn*tau*view(hm,1:j,1:j))
```
Which is better.

Reply:
this is becuase i'm a Julia noob. i instantiated the `F` outside the `while` loop because otherwise i would get "ERROR: UndefVarError: F not defined" error message.

Reply:
with

```julia
...
        # propagate using adaptive step size
        iter = 0
        while iter <= maxiter
            mx = mb + max(1,k1)
            F = expm!(tsgn*tau*view(hm, 1:mx, 1:mx))
...
```
i still get `ERROR: UndefVarError: F not defined`. i understand that this is because `F` is local to the scope of the `while` (["soft local scope"?](https://docs.julialang.org/en/stable/manual/variables-and-scoping/#Soft-Local-Scope-1)).

however, i can use `global F = ...`. do you like this solution or you propose to use a different workaround?

Reply:
Globals are almost never a good idea. This issue was one of the reasons I rewrote `expmv` getting rid of `k1` etc.

Reply:
I think you can just write `local F` outside of the loop, without further defining it, just declaring it.